subsys/fs/littlefs_fs.c

```diff
@@ -595,8 +613,42 @@ static int littlefs_mount(struct fs_mount_t *mountp)
 	dev = flash_area_get_device(fs->area);
 	if (dev == NULL) {
 		LOG_ERR("can't get flash device: %s", log_strdup(fs->area->fa_dev_name));
-		ret = -ENODEV;
-		goto out;
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+static int littlefs_mount(struct fs_mount_t *mountp)
+{
+	int ret = 0;
+	struct fs_littlefs *fs = mountp->fs_data;
+
+	LOG_INF("LittleFS version %u.%u, disk version %u.%u",
+		LFS_VERSION_MAJOR, LFS_VERSION_MINOR,
+		LFS_DISK_VERSION_MAJOR, LFS_DISK_VERSION_MINOR);
+
+	if (littlefs_on_blkdev(mountp)) {
+#ifdef CONFIG_FS_LITTLEFS_BLK_DEV
+		if (disk_access_init(mountp->storage_dev) != 0) {
+			LOG_ERR("Storage init ERROR!");
+			return -ENODEV;
+		}
+#endif
+	} else {
+		if (fs->area) {
+			return -EBUSY;
+		}
+	}
+
+	/* Create and take mutex. */
+	k_mutex_init(&fs->mutex);
+	fs_lock(fs);
+
+	if (!littlefs_on_blkdev(mountp)) {
```

Comment:
This thing gets called, now, second time in a row (and two more call are to come), I think that it is worth spending some stack variable to cache a result of the first call: this is doing `strcmp` every time.

Reply:
I also thought about directly check for each char - for example:
`if (mountp->storage_dev[0] == 'S' && mountp->storage_dev[1] == 'D' && ....)`

In that way the check can be hardcoded without the need to call `strcmp`.

However, I do agree that, the variable with cached value may be better to use.

Reply:
Cache the call into some bool variable,  it is not like the result of the call gets different every time and needs to be verified; in worst case scenario when optimizations somehow fail you will have all the costs of performing the call repeating through the function.

Reply:
@de-nordic, shouldn't this be replaced by a check of your recently introduced (and merged) flag ?

Reply:
@de-nordic This is already changed in the new sent today.

Reply:
Does not seem to be changed, as visible via commit https://github.com/zephyrproject-rtos/zephyr/pull/40704/commits/974587112129aed7bf67391eaeec013254a0f46e in this PR
https://github.com/zephyrproject-rtos/zephyr/blob/974587112129aed7bf67391eaeec013254a0f46e/subsys/fs/littlefs_fs.c#L627
https://github.com/zephyrproject-rtos/zephyr/blob/974587112129aed7bf67391eaeec013254a0f46e/subsys/fs/littlefs_fs.c#L644
https://github.com/zephyrproject-rtos/zephyr/blob/974587112129aed7bf67391eaeec013254a0f46e/subsys/fs/littlefs_fs.c#L679
https://github.com/zephyrproject-rtos/zephyr/blob/974587112129aed7bf67391eaeec013254a0f46e/subsys/fs/littlefs_fs.c#L724
https://github.com/zephyrproject-rtos/zephyr/blob/974587112129aed7bf67391eaeec013254a0f46e/subsys/fs/littlefs_fs.c#L761

Reply:
I thought that we have agreed that I shall use:
`if (littlefs_on_blkdev(mountp))` and then in the littlefs.h:

```
static inline bool littlefs_on_blkdev(struct fs_mount_t *mountp)
{
    if (IS_ENABLED(CONFIG_FS_LITTLEFS_BLK_DEV) && mountp->flags & FS_MOUNT_FLAG_USE_DISK_ACCESS)) {
           return true;
    }

    return false;
}
```

In that way the call to `littlefs_on_blkdev` will evaluate to an empty function and will be removed by LTO optimization.

Do you require to use the following pattern:
` if (IS_ENABLED(CONFIG_FS_LITTLEFS_BLK_DEV)  && littlefs_on_blkdev(mountp))` in the `littlefs_fs.c` mount function? 
So the `IS_ENABLED(CONFIG_FS_LITTLEFS_BLK_DEV)` will be used twice?

IMHO the approach proposed in this patch set is concise and neat.

Reply:
No, the function is fine internally.

I just suggest:
```
bool block_dev = littlefs_on_blkdev(mountp);
```
and use block_dev instead of invoking the function again.

Build the sample
```
west build -d mt_size -b nucleo_h743zi samples/subsys/fs/littlefs_blk/ -- -DCONFIG_SIZE_OPTIMIZATIONS=y
west build -d mt_speed -b nucleo_h743zi samples/subsys/fs/littlefs_blk/ -- -DCONFIG_SPEED_OPTIMIZATIONS=y
```
Change the code and build again, you will notice that a few bytes will be shaved off the binary with result of `littlefs_on_blkdev` cached, in mount.
At least experiment with this.

Reply:
Ok, I will add caching of the `littlefs_on_blkdev(mountp)` when it is called more than once.

The savings is neglectable -> ~20 flash bytes in program which has ~64KiB.

Reply:
> The savings is neglectable -> ~20 flash bytes in program which has ~64KiB.

It is only when the code that would require the saving to be hacky and ugly. You just gave 20 bytes of flesh to user app.